neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,493 @@
+    // `ShardedFunctionCallAction`.
+    fn call_sharded_contract(
+        // The account id of the account to call
+        account_id: AccountId,
+        // Which sharded contract code on the destination account to call
+        sharded_contract_account_id: AccountId,
+        // In between a sender sending the message and the receiver executing
+        // it, the contract code could have been upgraded.  The two fields below
+        // allow the sender to specify which versions the receiver can be in
+        // when called.  If the receiver's version is outside the specified
+        // range, then the call is rejected.
+        minimum_version: Option<u64>,
+        maximum_version: Option<u64>,
+        // These two arguments are a simplification of the existing args
+        // for the normal cross contract call.
+        function: &str,
+        args: Balance,
+    );
+}
+
+enum ShardedContractInfo {
+    Immutable { 
+        code_hash: CryptoHash
+    },
+    Mutable {
+        /// Account id of the account where the sharded contract code is deployed.
+        account_id: AccountId,
+        /// How many times the the above account has deployed a sharded contract
+        // code on itself.
+        version: u64,
+    },
+}
+
+// A smart contract function that can be used to initiate a FT transfer.
+fn send_tokens(amount: Balance, receiver: AccountId) {
+    // Since only the owner of this account is allowed to transfer funds out of
+    // it, ensure that the caller of this function is the owner of this account.
+    let my_account_id = HostFunctions::current_account_id();
+    let msg_sender = HostFunctions::predecessor_account_id();
+    assert_eq!(my_account_id, msg_sender);
+
+    // Update the account balance
+    let mut my_balance = HostFunctions::storage_read("balance");
+    assert!(my_balance >= amount);
+    my_balance -= amount;
+    HostFunctions::storage_write("balance", my_balance);
+
+    // Call the receiver's account to receive the tokens.
+    let my_sharded_contract_info = HostFunctions::current_sharded_contract_info();
+    HostFunctions::call_sharded_contract(
+        receiver,
+        my_sharded_contract_info.account_id,
+        // The sharded contract code on the receiver should be at the same
+        // precise version as the current account.
+        Some(my_sharded_contract_info.version),
+        Some(my_sharded_contract_info.version),
+        "receive_tokens",
+        amount,
+    );
+
+    // TODO: Should we emit an ft_transfer events here?  Or at the receiver?  Both?
+
+    // This pseudocode assumes that `receive_tokens()` always succeeds.  In a
+    // more complete version, if `receive_tokens()` fails, then this function
+    // should undo the balance decrement above to ensure that tokens are not
+    // lost.
+}
+
+// A smart contract function used to receive FT from a transfer.
+fn receive_tokens(amount: Balance) {
```

Comment:
I think I found a problem here.  What if the account owner end user calls this function directly?  They will be using the `ShardedFunctionCallAction` and at least how we have specified the checks here, they will pass.  We need to think about what `current_sharded_contract_info()` returns if the caller is an end user.  

CC: @jakmeier

Reply:
If they call it directly, wouldn't `predecessor_sharded_contract_info` return None? (Or panic)